api/openapi-spec/v3/api__v1_openapi.json

```diff
@@ -23,7 +27,7 @@
         },
         "type": "object"
       },
-      "io.k8s.api.authentication.v1.TokenRequest": {
+      "authentication.v1.TokenRequest": {
```

Comment:
am not convinced that we actually want to drop parts of the group name. Maybe it's a misunderstanding of the goal. Having the full group name here is fine, but the package name is not. In this example, we want `io.k8s.authentication.v1.TokenRequest`. Note that we drop the `api` which came from the package name.

Compare https://github.com/kubernetes/api/blob/master/authentication/v1/register.go#L26

Reply:
I agree. This is an unfortunate consequence of using path manipulation to determine the definition name. It seems though that we want the GroupName that is registered to the schema to be used.

I like that better, however I am unsure what to do in the scenario where there are possibly multiple group names associated with the package path, as the code around `GetDefinitionName` makes seem expected:

```
func (d *DefinitionNamer) GetDefinitionNamev3(name string) (string, spec.Extensions) {
	if groupVersionKinds, ok := d.typeGroupVersionKinds[name]; ok {
		return friendlyNamev3(name), spec.Extensions{
			extensionGVK: groupVersionKinds.JSON(),
		}
	}
	return friendlyNamev3(name), nil
}
```

In the above, `groupVersionKinds` is an array. For the case of at least 2 elements, I wonder how to determine which element is appropriate to use for the definition name.

Reply:
wouldn't there be two (or more) definitions in that case of multiple possible kinds, and they are referenced from the endpoint spec?

Reply:
this extension might be different. That's kube's "invention". Maybe we didn't prepare for multiple GVKs. Could well be.

Reply:
In the current implementation I have done as we discussed above, and changed the implementation to look up the type in the scheme to find the GVK. If no GVK is found, it reverts back to the original implementation. If multiple GVKs are found, the first one is used (the GVKs are sorted).

```
func (d *DefinitionNamer) GetDefinitionNameV3(name string) (string, spec.Extensions) {
	if groupVersionKinds, ok := d.typeGroupVersionKinds[name]; ok &&
		len(groupVersionKinds) > 0 {
		gvk := groupVersionKinds[0]
		group := gvk.Group

		// If the group is an empty string then it is part of the core apigroup
		if len(group) == 0 {
			group = "core"
		}

		return friendlyName(fmt.Sprintf("%s/%s.%s", group, gvk.Version, gvk.Kind)),
			spec.Extensions{
				extensionGVK: groupVersionKinds.JSON(),
			}
	}

	return friendlyName(name), nil
}
```

The above implementation works generally. In fact, I added your example of `io.k8s.authentication.v1.TokenRequest` as a test case. However, there are some issues I can identify:

1. The conflicting GVK-vs-type path issue I raised in the other comment chain.
2. Since this implementation just takes the alphabetically first GVK, if a new GVK were to be added to the type that is now alphabetically first, the type's name will change during the next spec generation. Is this acceptable? If it is not, I'm not sure removing the package path from the definition names is something that is possible.